homeassistant/components/tplink/sensor.py

```diff
@@ -34,23 +38,73 @@
 from .coordinator import TPLinkDataUpdateCoordinator
 from .entity import CoordinatedTPLinkEntity
 
+_LOGGER = logging.getLogger(__name__)
+
+
+AcceptedValueFnReturnValues = float | datetime.datetime | None
+
 
 @dataclass
 class TPLinkSensorEntityDescription(SensorEntityDescription):
     """Describes TPLink sensor entity."""
 
-    emeter_attr: str | None = None
+    attribute_name: str | None = None
+    value_fn: Callable[
+        [SmartDevice, TPLinkSensorEntityDescription], AcceptedValueFnReturnValues
+    ] | None = None
     precision: int | None = None
 
 
-ENERGY_SENSORS: tuple[TPLinkSensorEntityDescription, ...] = (
+def async_handle_emeter_attr(
```

Comment:
This was moved from below here to allow passing it as `value_fn` for the sensor definitions.